hivemq-edge/src/main/java/com/hivemq/protocols/DelegatingCompletableFuture.java

```diff
@@ -0,0 +1,699 @@
+/*
+ *  Copyright 2019-present HiveMQ GmbH
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package com.hivemq.protocols;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.annotations.UnknownNullability;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+public final class DelegatingCompletableFuture<T> extends CompletableFuture<T> {
+
+    private volatile @Nullable CompletableFuture<T> future;
+
+    public DelegatingCompletableFuture() {
+        future = null;
+    }
+
+    public boolean isEmpty() {
+        return future == null;
+    }
+
+    public @Nullable CompletableFuture<T> getFuture() {
+        return future;
+    }
+
+    public void setFuture(@Nullable final CompletableFuture<T> future) {
+        this.future = future;
+    }
+
+    @Override
+    public T get(final long timeout, final @NotNull TimeUnit unit)
+            throws InterruptedException, ExecutionException, TimeoutException {
+        final CompletableFuture<T> future = this.future;
+        if (future == null) {
+            return null;
```

Comment:
I am worried about all the return null within the future. All of these methods are actually treated as not ever returning a null in the original implementation.

Reply:
You are right. I think there are 2 options:

1. Throw an `ExecutionException`.
2. Sleep a while and wait for `future` to be not null periodically.

(1) doesn't seem to be friendly to the consumers. (2) seems to be a good solution. What do you think? Or you have any other solutions?

Reply:
I suppose the current solution and my proposed solution are eventually the same, as there is a spin lock waiting for the future to be set. It's more on where the complexity lies. I'm fine with keeping the current solution.

Reply:
Ok, then I will go back to fix the remaining issue.